hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java

```diff
@@ -485,6 +545,7 @@ public SerializerAdapter serializerFor(final Object object) {
         // 3-Custom registered types by user
         // 4-JDK serialization ( Serializable and Externalizable ) if a global serializer with Java serialization not registered
         // 5-Global serializer if registered by user
+        // 6-A good candidate to be serialized via compact serializer
```

Comment:
CompactSerializer is chosen in the second step if user explicitly registers it, similar to Portable. 

Other than that, we are trying to make use of it as last resort. If we put it before JDK, an Object implementing `Serializable` can not be serialized via JDK at all. 

This is similar to GlobalSerializer. IF we put compact before GlobalSerializer, then even if a user registers a globalSerializer all will be serialized via Compact. 

This is because of the zero-config nature of Compact Serializable objects. There is nothing on the class to decide if this is compact serializable or not. But not that when user configures it we make it priority in the second step.

Reply:
Isn't it better to use compact rather than Java when it is Serializable?

Reply:
Well. Java can serialize every `Serializable`, compact can not.

Reply:
We will discuss this after the Beta resolving for now.